### ISSUE TITLE
Compile failure with  `bytestring < 0.10`

### ISSUE THREAD

**Issue by hvr** [aelve/text-all#1]
```
Configuring library for text-all-0.4.1.0..
Preprocessing library for text-all-0.4.1.0..
Building library for text-all-0.4.1.0..
[1 of 2] Compiling Data.Text.Lazy.All ( lib/Data/Text/Lazy/All.hs, /tmp/matrix-worker/1499610637/dist-newstyle/build/x86_64-linux/ghc-7.4.2/text-all-0.4.1.0/build/Data/Text/Lazy/All.o )
[2 of 2] Compiling Data.Text.All    ( lib/Data/Text/All.hs, /tmp/matrix-worker/1499610637/dist-newstyle/build/x86_64-linux/ghc-7.4.2/text-all-0.4.1.0/build/Data/Text/All.o )

lib/Data/Text/All.hs:147:47:
    Not in scope: `BSL.toStrict'
    Perhaps you meant `TL.toStrict' (imported from Data.Text.Lazy)
```

i.e. you need to set version bounds on `bytestring`, e.g.  `bytestring >= 0.10 && < 0.11`

See also https://matrix.hackage.haskell.org/package/text-all

**Comment by neongreen** [comment]
Thanks! I've released text-all-0.4.1.1 with a lower bound on bytestring (and deprecated 0.4.1.0).

**Comment by hvr** [comment]
@neongreen sorry, I should have explained in more detail how Hackage works :-)

all you needed to do would have been to correct the meta-data over at 

- https://hackage.haskell.org/package/text-all-0.4.1.0/text-all.cabal/edit

which I've done now for you, see

 - https://hackage.haskell.org/package/text-all-0.4.1.0/revisions/

Uploading a new release when it can be corrected via a revision is discouraged for various technical reasons; and when it comes to making bounds tighter, you *must* perform a meta-data revision for the affected releases anyway, as otherwise the problem isn't fixed due to the way cabal's solver works.

PS: Deprecation doesn't mean that the solver isn't allowed to use `text-all-0.4.1.0`. It has a much weaker effect.

**Comment by neongreen** [comment]
> Uploading a new release when it can be corrected via a revision is discouraged for various technical reasons

Oh, okay, thanks. I didn't know that. Can I read about those technical reasons somewhere?

(Other than cabal's solver, of course – I knew about it but I thought that here choosing not to edit the metadata was justified because editing metadata annoys lots of people and GHC 7.4 is a rare beast anyway.)

**Comment by hvr** [comment]
> Can I read about those technical reasons somewhere?

Some of them are explained/hinted at in https://github.com/haskell-infra/hackage-trustees/blob/master/cookbook.md#best-practice-for-managing-meta-data and others in http://pvp.haskell.org/faq

I realise now that we don't have a real focused FAQ about this yet. I should write something more concise up when I have time. But if after reading the links above you still have questions, please feel free to ask me!

> because editing metadata annoys lots of people

Can you elaborate on this? I didn't have the impression that it annoys "lots of people", in the contrary, correcting metadata is essential for Hackage/Cabal users; without accurate meta-data the cabal solver would soon become completely useless.

**Comment by neongreen** [comment]
> I didn't have the impression that it annoys "lots of people"

There definitely seems to be a vocal minority pushing for “revisions are a misfeature” – e.g. see https://github.com/haskell-infra/hackage-trustees/issues/106#issuecomment-311425431:

> Revisions make life harder for everyone down-stream of Hackage except cabal-install and I'm not going to willingly participate in that.

> Almost every other ecosystem understands the importance of immutable packages. How absurd is it that a clique of _Haskell_ developers think they need mutable packages?

I don't think that revisions are a misfeature (at least not until we devise a better mechanism for preventing breakage), but at the same time it seemed to me that if in some particular case I could avoid creating a revision, I probably should – i.e. I thought that revisions are for the cases where the maintainer couldn't be reached, not for cases where an active maintainer wants to bump an upper bound or tighten incorrect bounds.

I still don't understand what are the technical reasons, though. Is it only about minimising solver's work? If so, do I understand correctly that if we had a way to tell cabal-install `please don't install this version under any circumstances`, the “New package releases ought to be avoided when a meta-data revision would suffice” recommendation from the cookbook would no longer apply?

**Comment by neongreen** [comment]
(NB. I realize now that it's actually quite questionable that the existence of a vocal minority implies that there's also “a lot” of people who aren't vocal but have the same preference. So, let's pretend that instead of “annoys lots of people” I had said “annoys some people a lot” without specifying how many such people there are :)

**Comment by hvr** [comment]
As to the question if this still would be a problem if we had support for blacklisting releases (similar to the deprecation feature): This would still result in a new version, which then means less sharing in the nix-style store, unneeded recompilation, but more importantly, previously fine install-plans would be suddenly invalidated and this would make reproducible builds harder as freeze-files (or exact-version `build-depends`!) would break much easier.

**Comment by neongreen** [comment]
Ah. Okay, makes sense, thanks.
